src/Wizard.js

```diff
@@ -1006,24 +1005,18 @@ export default class Wizard extends Webform {
     }
 
     // If the pages change, need to redraw the header.
-    let currentPanels;
-    let panels;
+    const currentPanels = this.pages;
+    // calling this.establishPages() updates/mutates this.pages to be the current pages
+    this.establishPages();
+    const newPanels = this.pages;
     const currentNextPage = this.currentNextPage;
-    if (this.hasExtraPages) {
-      currentPanels = this.pages.map(page => page.component.key);
-      this.establishPages();
-      panels = this.pages.map(page => page.component.key);
+    const panelsUpdated = !_.isEqual(newPanels, currentPanels);
+    
+    if (this.currentPanel?.id && this.pages.length && (!this.hasSubWizards || (this.hasSubWizards && panelsUpdated))) {
+      const newIndex = this.pages.findIndex(page => page.id === this.currentPanel.id);
+      if (newIndex !== -1) this.setPage(newIndex);
     }
-    else {
-      currentPanels = this.currentPanels || this.pages.map(page => page.component.key);
```

Comment:
`this.currentPanels` was only referenced within `onChange` and seems like an unnecessary piece of state given that the current pages/panels can be accessed via `this.pages`. All the tests pass happily without it.